### PR TITLE
Limit neighbors

### PR DESCRIPTION
**Pull request by ghabbenjansen**
Implements #236

### CODE REVIEW

**Comment by vandenheuvel** [comment]
retest this please

**Comment by ghabbenjansen** [comment]
`get_graph` still has too many local variables, but this method will be changed when converting the code to use twisted, so I will not force a solution when the code will be changed anyways

File: Tribler/community/triblerchain/community.py [hunk 1]
```diff
@@ -183,76 +183,112 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
+        # Limit the number of higher-level neighbors per node in the graph
+        for _ in range(0, neighbor_level):
+            for node in this_level:
+                return_nodes.append({"public_key": node, "total_up": nodes[node]["total_up"],
+                                     "total_down": nodes[node]["total_down"],
+                                     "total_neighbors": nodes[node]["total_neighbors"]})
+                nodes_visited.add(node)
+                num_edges = 0
+                for edge in edges[node]:
+                    if edge[0] in this_level:   # Free edge, opposite node in same level
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/community.py]: Please don't append comments to lines.

File: Tribler/community/triblerchain/community.py [hunk 2]
```diff
@@ -183,76 +183,112 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
+        # Check if the list of edges is empty
+        if not edges:
+            return_nodes.append({"public_key": public_key, "total_up": nodes[public_key]["total_up"],
+                                 "total_down": nodes[public_key]["total_down"],
+                                 "total_neighbors": nodes[public_key]["total_neighbors"]})
+            return_nodes[0]["score"] = calculate_score(return_nodes[0])
+            return return_nodes, return_edges
+
+        # Limit the number of higher-level neighbors per node in the graph
+        for _ in range(0, neighbor_level):
```
Comment (MaxVanDeursen) [inline Tribler/community/triblerchain/community.py]: `range(0, neighbor_level)` can be changed into `range(neighbor_level)`

File: Tribler/community/triblerchain/community.py [hunk 3]
```diff
@@ -183,76 +183,112 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
+                        next_level.add((edge[0]))
+                        return_edges.append({"from": node, "to": edge[0], "amount": edge[1]})
+                        return_edges.append({"from": edge[0], "to": node, "amount": edge[2]})
+                        if num_edges == max_neighbors:  # Stop adding edges after limit is reached
+                            break
+                    # Else the node is in a lower level and thus the edge is not shown
+            this_level = next_level
+            next_level = set()
+
+        for node in this_level:     # Add last level nodes
```
Comment (MaxVanDeursen) [inline Tribler/community/triblerchain/community.py]: Please add the comment to the line above.

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 4]
```diff
@@ -300,7 +300,9 @@ def render_GET(self, request):
 
         neighbor_level = self.get_neighbor_level(request.args)
 
-        nodes, edges = tribler_chain_community.get_graph(focus_node, neighbor_level)
+        max_neighbors = self.get_max_neighbors(request.args)
```
Comment (MaxVanDeursen) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: Please add documentation for the newly added parameter.

**Comment by MaxVanDeursen** [comment]
retest this please

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 5]
```diff
@@ -355,3 +374,35 @@ def get_neighbor_level(arguments):
...
+
+        The default is [user_node].
+        :param arguments: the arguments supplied with the HTTP request
+        :param user_node: the public key of the user
+        :return: list of mandatory nodes
+        """
+        mandatory_nodes = {}
+        if "mandatory_nodes" in arguments and arguments["mandatory_nodes"][0] != "undefined":
+            mandatory_nodes = arguments["mandatory_nodes"][0].split(",")
+        if user_node not in mandatory_nodes:
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: It's probably better to add the user node in the GUI.

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 6]
```diff
@@ -355,3 +374,35 @@ def get_neighbor_level(arguments):
...
+            mandatory_nodes.append(user_node)
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: You are calling `append` on a set/dict.

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 7]
```diff
@@ -355,3 +374,35 @@ def get_neighbor_level(arguments):
...
+    def get_mandatory_nodes(arguments, user_node):
+        """
+        Get the list of mandatory nodes.
+
+        The default is [user_node].
+        :param arguments: the arguments supplied with the HTTP request
+        :param user_node: the public key of the user
+        :return: list of mandatory nodes
+        """
+        mandatory_nodes = {}
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: It is ambiguous whether this is the `set` or `dict` constructor.

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 8]
```diff
@@ -355,3 +374,35 @@ def get_neighbor_level(arguments):
...
+    @staticmethod
+    def get_max_neighbors(arguments):
+        """
+        Get the maximum amount of higher level neighbors for one node.
+
+        The default maximum is 8.
+        :param arguments: the arguments supplied with the HTTP request
+        :return: maximal number of higher level neighbors per node
+        """
+        max_neighbors = 8
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: Extract this default.

File: Tribler/community/triblerchain/community.py [hunk 9]
```diff
@@ -183,76 +183,116 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
 
-        return_nodes = nodes.values()
+        return_nodes = []
         return_edges = []
 
+        nodes_visited = set()
+        this_level = set()
+        this_level.add(public_key)
+        # If the focus is in the list, always add it to the view
+        # if own_node in nodes:
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/community.py]: Remove this commented code.

File: Tribler/community/triblerchain/community.py [hunk 10]
```diff
@@ -183,76 +183,116 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
+        if not edges:
+            return_nodes.append({"public_key": public_key, "total_up": nodes[public_key]["total_up"],
+                                 "total_down": nodes[public_key]["total_down"],
+                                 "total_neighbors": nodes[public_key]["total_neighbors"]})
+            return_nodes[0]["score"] = calculate_score(return_nodes[0])
+            return return_nodes, return_edges
+
+        # Limit the number of higher-level neighbors per node in the graph
+        for current_level in range(neighbor_level + 1):
+            for node in this_level:
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/community.py]: `this_level` contains only 1 value.

File: TriblerGUI/widgets/network_explorer/index.html [hunk 11]
```diff
@@ -10,7 +10,7 @@
 <body>
 
 <!-- GRAPH -->
-<svg id="graph" width="600" height="600" viewBox="0 0 600 600"></svg>
+<svg id="graph" width="600" height="600" viewBox="0 0 600 600" preserveAspectRatio="xMinYMid meet"></svg>
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_explorer/index.html]: Why are you making GUI changes in this pull request?

File: TriblerGUI/widgets/network_explorer/js/radial_view/radial_navigation.js [hunk 12]
```diff
@@ -14,24 +14,33 @@ function RadialNavigation(requestMethod) {
     self.pending = false;
     self.current_pk = null;
     self.history = [];
+    self.mandatory_nodes = [];
     self.squash_history = true;
     self.neighbor_level = 1;
+    self.max_neighbors = 4;
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_explorer/js/radial_view/radial_navigation.js]: We could put these values in a config.

File: Tribler/community/triblerchain/community.py [hunk 13]
```diff
@@ -183,76 +183,116 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
...
+                edges[from_pk] = []
+            edges[from_pk].append((to_pk, amount_up, amount_down))
+            if to_pk not in edges:
+                edges[to_pk] = []
+            edges[to_pk].append((from_pk, amount_down, amount_up))
+
+        return nodes, edges
+
+    @blocking_call_on_reactor_thread
+    def get_graph(self, public_key, neighbor_level, max_neighbors, mandatory_nodes):
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/community.py]: This method is too complex.

File: Tribler/community/triblerchain/community.py (hunk 10 above)
Comment (MaxVanDeursen) [inline Tribler/community/triblerchain/community.py]: `this_level` is updated each iteration. It only contains 1 value the first iteration (unless the chosen node only has 1 member of course).

**Comment by vandenheuvel** [comment]
retest this please

**Pull request merged**